core/src/main/java/org/eclipse/hono/config/FileFormat.java

```diff
@@ -0,0 +1,94 @@
+/**
+ * Copyright (c) 2017 Red Hat Inc and others.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Red Hat Inc - initial creation
+ */
+
+package org.eclipse.hono.config;
+
+import static java.util.Optional.empty;
+import static java.util.Optional.of;
+import static java.util.Optional.ofNullable;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+public enum FileFormat {
+    /**
+     * PEM encoded PKCS#1 or PKCS#8
+     */
+    PEM,
+    /**
+     * Java Key Store
+     */
+    JKS,
+    /**
+     * PKCS#12 key store
+     */
+    PKCS12;
+
+    private static final Map<String, FileFormat> EXTENSIONS = new HashMap<>();
+
+    static {
+        EXTENSIONS.put("jks", JKS);
+
+        EXTENSIONS.put("pem", PEM);
+        EXTENSIONS.put("key", PEM);
+        EXTENSIONS.put("crt", PEM);
+
+        EXTENSIONS.put("p12", PKCS12);
+        EXTENSIONS.put("pfx", PKCS12);
+    }
+
+    /**
+     * Use the provided file format or detect.
+     * 
+     * If the file format is not provided the method will try to detect it by a call to {@link #detect(String)}.
+     * 
+     * @param format The provided format, may be {@code null}
+     * @param path The path to check, may be {@code null}, in which the result will be {@link Optional#empty()}.
+     * @return The provided or detected file format, may return {@link Optional#empty()} but never {@code null}
+     */
+    public static Optional<FileFormat> orDetect(final FileFormat format, final String path) {
+
+        if (path == null) {
+            return empty();
+        }
+
+        if (format != null) {
+            return of(format);
+        }
+
+        return detect(path);
+    }
+
+    /**
+     * Detect the file format based on the file name extension.
+     * 
+     * @param path The relative or absolute file name, may be {@code null}
+     * @return the detected file format, may return {@link Optional#empty()} but never {@code null}.
```

Comment:
My understanding is that returning `Optional.empty()` is supposed to support chaining another operation to the result. However, which operation would that be, given that `orDetect` already falls back to `detect` if no format is given?

Reply:
No .. that would be "completable futures". The idea of Optional is to never return null, but at the same time allowing to have "no result" as well.

Reply:
Please help me better understand, what is the reason for never wanting to return `null` in this context? Isn't that what we want to check for in the end anyway?

Reply:
The intention was to clearly indicate that there might be cases where no result can be provided by the method. `null` does that in a similar manner, but IMHO is less explicit and easily introduces errors. The only way to do that in Java 8 out of the box is using the Optional type.

Of course the same thing could be achieved by stating on the JavaDoc that the method might return `null` in some cases. I think this is more explicit. But I can re-write the PR to not using Optional if that helps.

Reply:
FMPOV returning `null` would be totally fine here (and indeed would be the most explicit way of saying *I have no idea what format this is*). It would also be more consistent with the rest of Hono's code (I cannot think of any method where we return an `Optional`). 

Client code needs to check for it anyway, i.e. it can either check `if (format.isPresent())` or do `if (format != null)`. I'd rather have the latter ...